### PR TITLE
draft pyramid workflow

### PR DESCRIPTION
**Pull request by jhamman**
cc @norlandrhagen

### CODE REVIEW

File: cmip6_downscaling/workflows/cmip6_pyramids.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
+import os
+
+import dask
+import datatree as dt
+import intake
+from adlfs import AzureBlobFileSystem
+from carbonplan_data.metadata import get_cf_global_attrs
+from carbonplan_data.utils import set_zarr_encoding
```
Comment (norlandrhagen) [inline cmip6_downscaling/workflows/cmip6_pyramids.py:8]: Should the `get_cf_global_attrs` and `set_zarr_encoding` be included in bcsd/gard flows as well?

File: cmip6_downscaling/workflows/cmip6_pyramids.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+# from prefect.executors import DaskExecutor
+# from dask_kubernetes import KubeCluster, make_pod_spec
+
+LEVELS = 2
+TIME_SLICE = slice('1950', '2100')
+# TIME_SLICE = slice('2010', '2020')
+PIXELS_PER_TILE = 128
+
+
+image = "carbonplan/cmip6-downscaling-prefect:2021.12.06"
```
Comment (norlandrhagen) [inline cmip6_downscaling/workflows/cmip6_pyramids.py:27]: A bunch of config settings now live in cmip6_downscaling/config/config.py if you want to use those.

File: cmip6_downscaling/workflows/cmip6_pyramids.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+# )
+# pod_spec.spec.containers[0].args.extend(['--resources', 'TASKSLOTS=1'])
+
+# executor = DaskExecutor(
+#     cluster_class=lambda: KubeCluster(pod_spec, deploy_mode='remote'),
+#     adapt_kwargs={"minimum": 1, "maximum": 2},
+# )
+
+
+def get_fs() -> AzureBlobFileSystem:
```
Comment (norlandrhagen) [inline cmip6_downscaling/workflows/cmip6_pyramids.py:66]: Could this be shared in utils.py?

File: cmip6_downscaling/workflows/cmip6_pyramids.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+@task(log_stdout=True)
+#  tags=['dask-resource:TASKSLOTS=1'], retry_delay=datetime.timedelta(seconds=30)
+def regrid(key: str, levels: int = LEVELS) -> None:
+
+    print(key, levels)
+
+    fs = get_fs()
+
+    with dask.config.set(scheduler='threads'):
```
Comment (norlandrhagen) [inline cmip6_downscaling/workflows/cmip6_pyramids.py:116]: Q: Is there a reason get_fs() is not in the dask context?

File: cmip6_downscaling/workflows/cmip6_pyramids.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+        cat = get_cat()
+
+        try:
+            ds = (
+                cat[key](storage_options={'account_name': 'cmip6downscaling'})
```
Comment (norlandrhagen) [inline cmip6_downscaling/workflows/cmip6_pyramids.py:122]: can the connection_string be passed here?

**Review by norlandrhagen — COMMENTED** [review]
Looks great!

File: cmip6_downscaling/workflows/cmip6_pyramids.py (hunk 1 above)
Comment (jhamman) [reply cmip6_downscaling/workflows/cmip6_pyramids.py:8]: yes please!

Comment (jhamman) [reply cmip6_downscaling/workflows/cmip6_pyramids.py:8]: In fact, we should specify that some of these fields are part of the intermediate and final dataset schemas

File: cmip6_downscaling/tasks/pyramid.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
+import dask
```
Comment (jhamman) [inline cmip6_downscaling/tasks/pyramid.py:1]: @norlandrhagen - can review this file only?

File: cmip6_downscaling/tasks/pyramid.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        )
+        dt[slevel].ds['time'].encoding['dtype'] = 'int32'
+        dt[slevel].ds['time_bnds'].encoding['dtype'] = 'int32'
+    dt.ds.attrs.update(**get_cf_global_attrs())
+    return dt
+
+
+@task(log_stdout=True)
+# Question: How should we
+# tags=['dask-resource:TASKSLOTS=1']
```
Comment (norlandrhagen) [inline cmip6_downscaling/tasks/pyramid.py]: It seems like this tasks called `pyramid_regrid`, which uses xesmf. Do you want to keep the tags['dask-resource:TASKSLOTS=1'] in?

File: cmip6_downscaling/tasks/pyramid.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    '''
+
+    with dask.config.set(scheduler='threads'):
+
+        ds.coords['date_str'] = ds['time'].dt.strftime('%Y-%m-%d').astype('S10')
+
+        ds = ds.pipe(_load_coords)
+
+        mapper = fsspec.get_mapper(uri)
+        # Question: Is this needed or will prefect handle this for us?
```
Comment (norlandrhagen) [inline cmip6_downscaling/tasks/pyramid.py:77]: You mean will prefect handle skip writing if store exists or?

File: cmip6_downscaling/tasks/pyramid.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+            return
+
+        # create
+        dt = pyramid_regrid(ds, target_pyramid=None, levels=levels)
+
+        # postprocess
+        dt = postprocess(dt, levels)
+
+        # write to uri
+        dt.to_zarr(mapper, mode='w')
```
Comment (norlandrhagen) [inline cmip6_downscaling/tasks/pyramid.py:88]: Add consolidated=True?

File: cmip6_downscaling/tasks/pyramid.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    dt.DataTree
+        Updated data pyramid with metadata / encoding set
+    '''
+
+    for level in range(levels):
+        slevel = str(level)
+        dt.ds.attrs['multiscales'][0]['datasets'][level]['pixels_per_tile'] = PIXELS_PER_TILE
+
+        dt[slevel].ds = dt[slevel].ds.chunk(
+            {"x": PIXELS_PER_TILE, "y": PIXELS_PER_TILE, "time": 31}
```
[the remark this answers is not in the document]
Comment (jhamman) [reply cmip6_downscaling/tasks/pyramid.py:43]: great point, this should be changed.

File: cmip6_downscaling/tasks/pyramid.py (hunk 9 above)
Comment (jhamman) [reply cmip6_downscaling/tasks/pyramid.py:88]: The default behavior is in xarray is to consolidate metadata so this is not needed.

**Pull request merged**